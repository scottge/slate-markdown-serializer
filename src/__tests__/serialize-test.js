import MarkdownRenderer from '../renderer'
const Markdown = new MarkdownRenderer()

function reserialized(text) {
  return Markdown.serialize(Markdown.deserialize(text))
}

test('serializes paragraph', () => {
  const text = 'This is just a sentance'
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes one paragraph', () => {
  const text = `
This is the first sentance
This is the second sentance
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes two paragraphs', () => {
  const text = `
This is the first sentance

This is the second sentance
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes two paragraphs seperated by multiple blank lines', () => {
  const text = `
This is the first sentance


This is the second sentance
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes heading1', () => {
  const text = `# Heading`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes heading2', () => {
  const text = `## Heading`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes heading3', () => {
  const text = `### Heading`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes heading4', () => {
  const text = `#### Heading`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes heading5', () => {
  const text = `##### Heading`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes heading6', () => {
  const text = `###### Heading`
  expect(reserialized(text)).toMatchSnapshot()
})


test('serializes headings followed by newlines', () => {
  const text = `
a paragraph

## Heading

another paragraph
`
  expect(reserialized(text)).toMatchSnapshot()
})


test('serializes horizontal rule', () => {
  const text = `
---

a paragraph
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('bold mark', () => {
  const text = `**this is bold**`
  expect(reserialized(text)).toMatchSnapshot()
})

test('italic mark', () => {
  const text = `*this is italic* _this is italic too_`
  expect(reserialized(text)).toMatchSnapshot()
})

test('deleted mark', () => {
  const text = `~~this is strikethrough~~`
  expect(reserialized(text)).toMatchSnapshot()
})

test('inserted mark', () => {
  const text = `++inserted text++`
  expect(reserialized(text)).toMatchSnapshot()
})

test('underlined mark', () => {
  const text = `__underlined text__`
  expect(reserialized(text)).toMatchSnapshot()
})

test('code mark', () => {
  const text = '`const foo = 123;`'
  expect(reserialized(text)).toMatchSnapshot()
})

test('code mark with escaped characters', () => {
  const text = "`<script>alert('foo')</script>`"
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes quote', () => {
  const text = `
> this is a quote
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes quote followed by list with quote', () => {
  const text = `
> this is a quote
1. > this is a list item
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes quotes separated by newlines', () => {
  const text = `
> this is a quote

> this is a different quote
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes quote followed by newlines', () => {
  const text = `
> this is a quote

this is a paragraph
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes list items', () => {
  const text = `
- one
- two
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes nested list items', () => {
  const text = `
* one
* two
   * nested

next para`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes list surrounded by paragraphs', () => {
  const text = `
first paragraph

- list

second paragraph
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes indented list items', () => {
  const text = `
 - one
 - two
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes list items with marks', () => {
  const text = `
 - one **bold**
 - *italic* two
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes ordered list items', () => {
  const text = `
1. one
1. two
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes ordered list items with marks', () => {
  const text = `
1. one **bold**
1. *italic* two
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes ordered list items with different numbers', () => {
  const text = `
1. one
2. two
3. three
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes mixed list items', () => {
  const text = `
1. list

- another

1. different
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes tables', () => {
  const text = `
| Tables   |      Are      |  Cool |
|----------|:-------------:|------:|
| col 1 is |  left-aligned | $1600 |
| col 2 is |    centered   |   $12 |
| col 3 is | right-aligned |    $1 |
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes tables followed by newlines', () => {
  const text = `
| Tables   |      Are      |  Cool |
|----------|:-------------:|------:|
| col 1 is |  left-aligned | $1600 |

a new paragraph
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes ``` code fences', () => {
  const text = `
\`\`\`
const hello = 'world';
function() {
  return hello;
}
\`\`\`
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes ``` code fences with language', () => {
  const text = `
\`\`\`javascript
const hello = 'world';
function() {
  return hello;
}
\`\`\`
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('seserializes ``` code fences with escaped characters', () => {
  const text = `
\`\`\`
const hello = 'world';
function() {
  return hello;
}
\`\`\`
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes code followed by newlines', () => {
  const text = `
one sentance

\`\`\`
const hello = 'world';
function() {
  return hello;
}
\`\`\`

two sentance
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes ~~~ code fences', () => {
  const text = `
~~~
const hello = 'world';
function() {
  return hello;
}
~~~
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes indented code blocks', () => {
  const text = `
    const hello = 'world';
    function() {
      return hello;
    }
`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes hashtag', () => {
  const text = `this is a #hashtag example`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes hashtag ignoring dash', () => {
  const text = `dash should end #hashtag-dash`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes image', () => {
  const text = `![example](http://example.com/logo.png)`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes image with title', () => {
  const text = `![example](http://example.com/logo.png "image title")`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes link', () => {
  const text = `[google](http://google.com)`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes link within mark', () => {
  const text = `**[google](http://google.com)**`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes link with encoded characters', () => {
  const text = `[kibana](https://example.com/app/kibana#/discover?_g=%28refreshInterval:%28%27$$hashKey%27:%27object:1596%27,display:%2710%20seconds%27,pause:!f,section:1,value:10000%29,time:%28from:now-15m,mode:quick,to:now%29%29&_a=%28columns:!%28metadata.step,message,metadata.attempt_f,metadata.tries_f,metadata.error_class,metadata.url%29,index:%27logs-%27,interval:auto,query:%28query_string:%28analyze_wildcard:!t,query:%27metadata.at:%20Stepper*%27%29%29,sort:!%28time,desc%29%29)`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes link with percent symbol', () => {
  const text = `[kibana](https://example.com/app/kibana#/visualize/edit/Requests-%)`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes empty link', () => {
  const text = `[empty]()`
  expect(reserialized(text)).toMatchSnapshot()
})

test('serializes empty string', () => {
  expect(reserialized('')).toMatchSnapshot()
})

test('serializes whitespace string', () => {
  expect(reserialized('   ')).toMatchSnapshot()
})

test('serializes escaped blocks', () => {
  expect(reserialized('\\# text')).toMatchSnapshot()
  expect(reserialized('\\- text')).toMatchSnapshot()
  expect(reserialized('\\* text')).toMatchSnapshot()
})

test('serializes escaped marks', () => {
  expect(reserialized('this is \\*\\*not bold\\*\\*')).toMatchSnapshot()
  expect(reserialized('this is \\*not italic\\*')).toMatchSnapshot()
  expect(reserialized('this is \\[not\\]\\(a link\\)')).toMatchSnapshot()
  expect(reserialized('this is \\!\\[not\\]\\(an image\\)')).toMatchSnapshot()
})
