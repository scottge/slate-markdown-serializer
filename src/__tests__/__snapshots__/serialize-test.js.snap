// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bold mark 1`] = `"**this is bold**"`;

exports[`code mark 1`] = `"\`const foo = 123;\`"`;

exports[`code mark with escaped characters 1`] = `"\`<script>alert('foo')</script>\`"`;

exports[`deleted mark 1`] = `"~~this is strikethrough~~"`;

exports[`inserted mark 1`] = `"++inserted text++"`;

exports[`italic mark 1`] = `"_this is italic_ _this is italic too_"`;

exports[`serializes \`\`\` code fences 1`] = `
"\`\`\`
const hello = 'world';
function() {
  return hello;
}
\`\`\`"
`;

exports[`serializes \`\`\` code fences with language 1`] = `
"\`\`\`javascript
const hello = 'world';
function() {
  return hello;
}
\`\`\`"
`;

exports[`serializes ~~~ code fences 1`] = `
"\`\`\`
const hello = 'world';
function() {
  return hello;
}
\`\`\`"
`;

exports[`serializes code followed by newlines 1`] = `
"one sentance

\`\`\`
const hello = 'world';
function() {
  return hello;
}
\`\`\`

two sentance"
`;

exports[`serializes empty link 1`] = `"[empty]()"`;

exports[`serializes empty string 1`] = `""`;

exports[`serializes escaped blocks 1`] = `"\\\\# text"`;

exports[`serializes escaped blocks 2`] = `"\\\\- text"`;

exports[`serializes escaped blocks 3`] = `"\\\\* text"`;

exports[`serializes escaped marks 1`] = `"this is \\\\*\\\\*not bold\\\\*\\\\*"`;

exports[`serializes escaped marks 2`] = `"this is \\\\*not italic\\\\*"`;

exports[`serializes escaped marks 3`] = `"this is \\\\[not\\\\]\\\\(a link\\\\)"`;

exports[`serializes escaped marks 4`] = `"this is \\\\!\\\\[not\\\\]\\\\(an image\\\\)"`;

exports[`serializes hashtag 1`] = `"this is a #hashtag example"`;

exports[`serializes hashtag ignoring dash 1`] = `"dash should end #hashtag\\\\-dash"`;

exports[`serializes heading1 1`] = `"# Heading"`;

exports[`serializes heading2 1`] = `"## Heading"`;

exports[`serializes heading3 1`] = `"### Heading"`;

exports[`serializes heading4 1`] = `"#### Heading"`;

exports[`serializes heading5 1`] = `"##### Heading"`;

exports[`serializes heading6 1`] = `"###### Heading"`;

exports[`serializes headings followed by newlines 1`] = `
"a paragraph

## Heading

another paragraph"
`;

exports[`serializes horizontal rule 1`] = `
"---

a paragraph"
`;

exports[`serializes image 1`] = `"![example](http://example.com/logo.png)"`;

exports[`serializes image with title 1`] = `"![example](http://example.com/logo.png \\"image%20title\\")"`;

exports[`serializes indented code blocks 1`] = `
"\`\`\`
const hello = 'world';
function() {
  return hello;
}
\`\`\`"
`;

exports[`serializes indented list items 1`] = `
"* one
* two"
`;

exports[`serializes link 1`] = `"[google](http://google.com)"`;

exports[`serializes link with encoded characters 1`] = `"[kibana](https://example.com/app/kibana#/discover?_g=%28refreshInterval:%28%27$$hashKey%27:%27object:1596%27,display:%2710%20seconds%27,pause:!f,section:1,value:10000%29,time:%28from:now-15m,mode:quick,to:now%29%29&_a=%28columns:!%28metadata.step,message,metadata.attempt_f,metadata.tries_f,metadata.error_class,metadata.url%29,index:%27logs-%27,interval:auto,query:%28query_string:%28analyze_wildcard:!t,query:%27metadata.at:%20Stepper*%27%29%29,sort:!%28time,desc%29%29)"`;

exports[`serializes link with percent symbol 1`] = `"[kibana](https://example.com/app/kibana#/visualize/edit/Requests-%)"`;

exports[`serializes link within mark 1`] = `"[**google**](http://google.com)"`;

exports[`serializes list items 1`] = `
"* one
* two"
`;

exports[`serializes list items with marks 1`] = `
"* one **bold**
* _italic_ two"
`;

exports[`serializes list surrounded by paragraphs 1`] = `
"first paragraph

* list

second paragraph"
`;

exports[`serializes mixed list items 1`] = `
"1. list

* another

1. different"
`;

exports[`serializes nested list items 1`] = `
"* one
* two
  * nested

next para"
`;

exports[`serializes one paragraph 1`] = `
"This is the first sentance
This is the second sentance"
`;

exports[`serializes ordered list items 1`] = `
"1. one
2. two"
`;

exports[`serializes ordered list items with different numbers 1`] = `
"1. one
2. two
3. three"
`;

exports[`serializes ordered list items with marks 1`] = `
"1. one **bold**
2. _italic_ two"
`;

exports[`serializes paragraph 1`] = `"This is just a sentance"`;

exports[`serializes quote 1`] = `"> this is a quote"`;

exports[`serializes quote followed by list with quote 1`] = `
"> this is a quote

1. > this is a list item"
`;

exports[`serializes quote followed by newlines 1`] = `
"> this is a quote

this is a paragraph"
`;

exports[`serializes quotes separated by newlines 1`] = `
"> this is a quote

> this is a different quote"
`;

exports[`serializes tables 1`] = `
"| Tables | Are | Cool |
| --- |:---:| ---:|
| col 1 is | left\\\\-aligned | $1600 |
| col 2 is | centered | $12 |
| col 3 is | right\\\\-aligned | $1 |"
`;

exports[`serializes tables followed by newlines 1`] = `
"| Tables | Are | Cool |
| --- |:---:| ---:|
| col 1 is | left\\\\-aligned | $1600 |

a new paragraph"
`;

exports[`serializes two paragraphs 1`] = `
"This is the first sentance

This is the second sentance"
`;

exports[`serializes two paragraphs seperated by multiple blank lines 1`] = `
"This is the first sentance

This is the second sentance"
`;

exports[`serializes whitespace string 1`] = `""`;

exports[`seserializes \`\`\` code fences with escaped characters 1`] = `
"\`\`\`
const hello = 'world';
function() {
  return hello;
}
\`\`\`"
`;

exports[`underlined mark 1`] = `"__underlined text__"`;
